### DISCUSSION TITLE
Win - does? Doesn't? Some thoughts.

### DISCUSSION THREAD

**Discussion by terminator356** [muse-sequencer/muse#1138] [Team Posts]
Hi Robert. I saw you mentioned Win32 somewhere and committed a skeletal README file.
Just want to pipe up and share some thoughts.

A few years ago I experimented trying to build on Windows.
First, I tried a native build. Got about 80% through, fixing all manner of common things,
 until I got to things like pthreads. Brick wall.
So then I tried mingw and msys2. What a nightmare that was. No luck at all.
Recently, I returned to try once and for all to do it.
Do you know what happened?
In the middle of working, unexpectedly without warning, free Avast! antivirus
 popped up and said that some files from mingw were threats.
Whereupon it moved them to a non-existent **bogus** 'quarantine' folder,
 nowhere to be found and none of the recover buttons worked.
I suspect it simply trashed them. Online I found many people with the same complaint.
When I try to reinstall mingw Avast! immediately trashes it again.
So screw that! No way I'm going down that mingw path again.
**So I will never be satisfied until we have a true native Windows build.**
We are edging closer to that possibility.
Probably the most sticking point is realtime pthreads replacement.
There's likely something native we can use.
QThread simply does NOT cut it. Even at it's highest priority setting,
 my investigation showed it is NOT realtime. I pointed that out to Rui once,
 QThread being used in QTractor, in a somewhat 'alarmist' message to him.

**Comment by kybos** [comment]
Probably a stupid question, sorry: What exactly is the purpose of a Windows version of MusE?

**Comment by spamatica** [comment]
The number one reason would be to increase user-base, the windows platform
is huge and, sticking to open source, the selection is not that big. A
secondary reason would be to clean the code base of unnecessarily platform
dependent code.

I did a branch many years ago platform_independence where I actually got it
up and running (to some extent) on OSX before I ran out of steam. Before
that happened there were a few good things that came out of it, like
switching some linux-only libs to Qt equivalent code, both removing
dependencies and making the code more portable.

More recently there was a guy (I forget his name, there are messages in the
forum) who did a lot of work to get it up and running on Windows, I'm
guessing for mostly the same reasons.

As for the issue with QThreads, in Windows I'm pretty sure the priority
settings works correctly.
In the Linux case shouldn't we be able to #ifdef LINUX and change scheduler
manually for the relevant threads?


Den fre 13 dec. 2019 kl 07:32 skrev kybos <notifications@github.com>:

**Comment by kybos** [comment]
True, there are not many open source DAWs on Windows, but to my experience most Windows users don't care much if it's OSS or freeware or anything. And there is now even Sonar for free...
But no arguing here, there is probably no reason not to do it, provided the required effort is reasonable and it does not mess up the code with hundreds of those ugly preprocessor directives...

**Comment by spamatica** [comment]
Den fre 13 dec. 2019 kl 15:06 skrev kybos <notifications@github.com>:

> True, there are not many open source DAWs on Windows, but to my experience
> most Windows users don't care much if it's OSS or freeware or anything. And
> there is now even Sonar for free...
>
Most users on Windows do not care about open source that is true, but as
the user base is huge, the few percent that exist should still be quite a
lot of people.

> But no arguing here, there is probably no reason not to do it, provided
> the required effort is reasonable and it does not mess up the code with
> hundreds of those ugly preprocessor directives...
>
I don't think there is any risk for that, not more than is already
available. There are some patches already applied to the master branch for
building on Windows.  Haven't tried to it myself yet but if I recall
correctly muse can already be built but is not fully working.

I would say this isn't prioritized by any dev currently. But it's
interesting, I'd like to return to it myself, so far it hasn't been
interesting enough though.
